libyul/backends/evm/ControlFlowGraphBuilder.cpp

```diff
@@ -0,0 +1,471 @@
+	Dialect const& _dialect,
+	Block const& _block
+)
+{
+	auto result = std::make_unique<CFG>();
+	result->entry = &result->makeBlock();
+
+	ControlFlowGraphBuilder builder(*result, _analysisInfo, _dialect);
+	builder.m_currentBlock = result->entry;
+	builder(_block);
+
+	// Determine which blocks are reachable from the entry.
+	util::BreadthFirstSearch<CFG::BasicBlock*> reachabilityCheck{{result->entry}};
+	for (auto const& functionInfo: result->functionInfo | ranges::views::values)
+		reachabilityCheck.verticesToTraverse.emplace_back(functionInfo.entry);
+
+	reachabilityCheck.run([&](CFG::BasicBlock* _node, auto&& _addChild) {
+		visit(util::GenericVisitor{
+			[&](CFG::BasicBlock::Jump& _jump) {
+				_addChild(_jump.target);
+			},
+			[&](CFG::BasicBlock::ConditionalJump& _jump) {
+				_addChild(_jump.zero);
+				_addChild(_jump.nonZero);
+			},
+			[](CFG::BasicBlock::FunctionReturn&) {},
+			[](CFG::BasicBlock::Terminated&) {},
+			[](CFG::BasicBlock::MainExit&) {}
+		}, _node->exit);
+	});
+
+	// Remove all entries from unreachable nodes from the graph.
+	for (CFG::BasicBlock* node: reachabilityCheck.visited)
+		cxx20::erase_if(node->entries, [&](CFG::BasicBlock* entry) -> bool {
+			return !reachabilityCheck.visited.count(entry);
+		});
+
+	// TODO: It might be worthwhile to run some further simplifications on the graph itself here.
+	// E.g. if there is a jump to a node that has the jumping node as its only entry, the nodes can be fused, etc.
+
+	return result;
+}
+
+ControlFlowGraphBuilder::ControlFlowGraphBuilder(
+	CFG& _graph,
+	AsmAnalysisInfo const& _analysisInfo,
+	Dialect const& _dialect
+):
+	m_graph(_graph),
+	m_info(_analysisInfo),
+	m_dialect(_dialect)
+{
+}
+
+StackSlot ControlFlowGraphBuilder::operator()(Literal const& _literal)
+{
+	return LiteralSlot{valueOfLiteral(_literal), _literal.debugData};
+}
+
+StackSlot ControlFlowGraphBuilder::operator()(Identifier const& _identifier)
+{
+	return VariableSlot{lookupVariable(_identifier.name), _identifier.debugData};
+}
+
+StackSlot ControlFlowGraphBuilder::operator()(Expression const& _expression)
+{
+	return std::visit(*this, _expression);
+}
+
+StackSlot ControlFlowGraphBuilder::operator()(FunctionCall const& _call)
+{
+	CFG::Operation const& operation = visitFunctionCall(_call);
+	yulAssert(operation.output.size() == 1, "");
+	return operation.output.front();
+}
+
+void ControlFlowGraphBuilder::operator()(VariableDeclaration const& _varDecl)
+{
+	yulAssert(m_currentBlock, "");
+	auto declaredVariables = _varDecl.variables | ranges::views::transform([&](TypedName const& _var) {
+		return VariableSlot{lookupVariable(_var.name), _var.debugData};
+	}) | ranges::to<vector>;
+	Stack input;
+	if (_varDecl.value)
+		input = visitAssignmentRightHandSide(*_varDecl.value, declaredVariables.size());
+	else
+		input = ranges::views::iota(0u, _varDecl.variables.size()) | ranges::views::transform([&](size_t) {
+			return LiteralSlot{0, _varDecl.debugData};
+		}) | ranges::to<Stack>;
```

Comment:
Is this the same as
```suggestion
		input = Stack(_varDecl.variables.size()), LiteralSlot{0, _varDecl.debugData});
```
?

Reply:
Haha, yeah, it is :-).